File: app/src/main/java/io/shelfy/architecture/presentation/details/DetailsFragment.java
package io.shelfy.architecture.presentation.details;

import android.content.Intent;
import android.net.Uri;
import android.os.Bundle;
import android.view.View;
import android.widget.ImageView;
import android.widget.TextView;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;

import com.bumptech.glide.Glide;

import java.util.Objects;

import io.shelfy.architecture.R;
import io.shelfy.architecture.domain.entity.Movie;
import io.shelfy.architecture.presentation.common.BaseFragment;

public class DetailsFragment extends BaseFragment {

    private static final String MOVIE_ID = "movie_id";

    private int movieId = -1;
    private View btnTrailer;

    private MovieDetailsViewModel viewModel;
    private ImageView ivBackground;
    private ImageView ivPoster;
    private TextView tvTitle;
    private TextView tvReleaseDate;
    private TextView tvOverview;

    public static DetailsFragment newInstance(Movie movie) {
        final DetailsFragment detailsFragment = new DetailsFragment();
        Bundle bundle = new Bundle();
        bundle.putInt(MOVIE_ID, movie.getId());
        detailsFragment.setArguments(bundle);
        return detailsFragment;
    }

    public DetailsFragment() {
        super(R.layout.fragment_movie_details);
    }

    @Override
    public void onCreate(@Nullable Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);

        movieId = Objects.requireNonNull(getArguments()).getInt(MOVIE_ID);
    }

    @Override
    public void onViewCreated(@NonNull View view, @Nullable Bundle savedInstanceState) {
        super.onViewCreated(view, savedInstanceState);
        ivBackground = view.findViewById(R.id.ivBackground);
        ivPoster = view.findViewById(R.id.ivPoster);
        tvTitle = view.findViewById(R.id.tvTitle);
        tvReleaseDate = view.findViewById(R.id.tvReleasedDate);
        tvOverview = view.findViewById(R.id.tvOverview);
        btnTrailer = view.findViewById(R.id.btnTrailer);

        viewModel = fragmentComponent.provideViewModel(MovieDetailsViewModel.class);

        initViewModelObservers();

        btnTrailer.setOnClickListener(v -> viewModel.loadTrailer(movieId).observe(
                getViewLifecycleOwner(),
                movieVideo -> openMovieTrailer(movieVideo.getVideoUrl())));
    }

    private void openMovieTrailer(String url) {
        Intent intent = new Intent(Intent.ACTION_VIEW, Uri.parse(url));
        startActivity(intent);
    }

    private void initViewModelObservers() {
        viewModel.getMovie(movieId).observe(getViewLifecycleOwner(), movie -> {
            Glide.with(ivBackground)
                    .load(movie.getBackdropUrl())
                    .into(ivBackground);
            Glide.with(ivPoster)
                    .load(movie.getPosterUrl())
                    .into(ivPoster);
            tvTitle.setText(movie.getTitle());
            tvReleaseDate.setText(movie.getReleaseDate());
            tvOverview.setText(movie.getDescription());
        });
    }
}

File: app/src/main/java/io/shelfy/architecture/ui/DetailsFragment.kt
package io.shelfy.architecture.ui

import android.content.Intent
import android.net.Uri
import android.os.Bundle
import android.view.View
import android.widget.Toast
import androidx.core.os.bundleOf
import androidx.fragment.app.Fragment
import androidx.lifecycle.Observer
import androidx.lifecycle.ViewModelProvider
import io.shelfy.architecture.Dependencies
import io.shelfy.architecture.R
import io.shelfy.architecture.domain.model.Movie
import io.shelfy.architecture.presentation.MovieDetailsViewModel
import io.shelfy.architecture.ui.formatters.MovieDescriptionFormatter
import coil.api.load
import kotlinx.android.synthetic.main.fragment_movie_details.*

class DetailsFragment : Fragment(R.layout.fragment_movie_details) {

    companion object {

        private const val ARG_MOVIE = "arg.movie"

        fun newInstance(movie: Movie): DetailsFragment {
            return DetailsFragment().apply {
                arguments = bundleOf(ARG_MOVIE to movie)
            }
        }
    }

    private lateinit var viewModel: MovieDetailsViewModel

    private val descriptionFormatter = MovieDescriptionFormatter()

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)

        val movie: Movie = arguments?.getParcelable(ARG_MOVIE)
            ?: throw IllegalArgumentException("missing argument!")

        val viewModelFactory = Dependencies.createMovieDetailsViewModelFactory(movie)
        viewModel = ViewModelProvider(this, viewModelFactory)
            .get(MovieDetailsViewModel::class.java)
    }

    override fun onViewCreated(view: View, savedInstanceState: Bundle?) {
        super.onViewCreated(view, savedInstanceState)

        initViewModelObservers()

        btnTrailer.setOnClickListener {
            viewModel.loadTrailer()
        }
    }

    private fun initViewModelObservers() {
        viewModel.movieLiveData.observe(viewLifecycleOwner, Observer { movie ->
            with(movie) {
                ivBackground.load(backdropUrl)
                ivPoster.load(posterUrl)
                tvTitle.text = title
                tvReleasedDate.text = releaseDate
                tvOverview.text = descriptionFormatter.format(this)
            }
        })
        viewModel.movieTrailerLiveData.observe(viewLifecycleOwner, Observer { movieVideo ->
            openMovieTrailer(movieVideo.videoUrl)
        })
        viewModel.errorLiveData.observe(viewLifecycleOwner, Observer { message ->
            Toast.makeText(requireContext(), message, Toast.LENGTH_LONG).show()
        })
    }

    private fun openMovieTrailer(url: String) {
        val intent = Intent(Intent.ACTION_VIEW, Uri.parse(url))
        startActivity(intent)
    }
}
